### ISSUE TITLE
Does not work in the IDLE shell

### ISSUE THREAD

**Issue by mardukbp** [onelivesleft/PrettyErrors#23]
The shell of the IDLE editor shipped with Python supports colors. It would be great to add support for it.

**Comment by onelivesleft** [comment]
Have you got pretty_errors working with IDLE?  I just tried it: importing it didn't affect the output, and calling `replace_stderr()` caused the shell to restart.

**Comment by mardukbp** [comment]
In IDLE 3.8.2 the following works:

```py
import pretty_errors
from pretty_errors import replace_stderr

replace_stderr()
```

But IDLE does not recognize ANSI escape sequences. You would have to define a dictionary with ANSI colors and a dictionary with IDLE colors, and that should be enough.

**Comment by onelivesleft** [comment]
Try doing `raise KeyError` after the code you posted: for me (in 3.7) it restarts the shell without outputing the exception.

IDLE's shell doesn't support simple colour codes like an ANSI compatible shell.  Instead it replaces stdout/stderr with its own class, which adds a 'shell' class with a write method you can call with a type string.  For example, `sys.stderr.shell.write(text, 'KEYWORD') will use the KEYWORD colour (orange by default).  If it worked with simple colour codes you could just use `pretty_errors.config` to change all the prefixes, but sadly that isn't an option.

Basically, IDLE monkeys around with both `sys.excepthook` *and* `sys.stderr`; I'm not sure how viable it is trying to shim pretty_errors in there at the same time. If there exists (or someone were to make) an IDLE extension that exposes exception handling, that could be used. I'll have a go at adding a dummy `shell` class to my dummy `stderr` class, but I'm not optimistic.

**Comment by onelivesleft** [comment]
Having looked into it, and thought about it some more, I'm afraid not going to worry about this for pretty_errors.  Firstly, writing special case code for one specific shell might be worthwhile if the shell is popular, but IDLE's userbase is vanishingly small.  3 years ago it polled at 10%, whereas last year it was down to 2%:

![image](https://user-images.githubusercontent.com/17881109/93101765-98fbe000-f6a2-11ea-9faf-9b6cb4650f0d.png)

Secondly, the fix for this is not to special-case pretty-errors; it's to update IDLE to handle ansi codes.  This could be done either directly in the IDLE source, or by writing an extension for it.  Alternatively, while writing full ansi support would be useful for a lot of people, adding a handful of recognized colour tokens (the ones output by your pretty errors config) could be relatively straightforward, if you just wanted to make an extension for yourself.

Thank you for the suggestion, it is a good one: I agree it would be nice for pretty_errors to work with IDLE because IDLE ships with every python install, but it's not going to happen for the reasons stated above.  Sorry!

**Comment by onelivesleft** [comment]
N.B. I should have said: if you want to keep using it in your current setup you can clean up the output by calling `pretty_errors.mono()`, which will remove all the ansi codes.   If you want to highlight things using ascii art then use `pretty_errors.configure()` to set the colour prefixes to whatever you want.

**Comment by mardukbp** [comment]
You are right. IDLE should be modified, not pretty_errors. Thank you so much for writing this library. I always wanted to have readable stack traces to quickly see why a program crashed. Now I can :)

**Comment by onelivesleft** [comment]
You are welcome!
